metricflow/inference/rule/base.py

```diff
@@ -0,0 +1,139 @@
+
+
+@dataclass(frozen=True)
+class InferenceSignal:
+    """Encapsulates a piece of evidence about a column produced by an inference rule.
+
+    column: the target column for this signal
+    type: the type of the signal
+    reason: a human-readable string that explains why this signal was produced. It may
+        eventually reach the user's eyeballs.
+    confidence: the confidence that this signal is correct.
+    """
+
+    column: SqlColumn
+    type: InferenceSignalType
+    reason: str
+    confidence: InferenceSignalConfidence
+
+
+RequiredContextsType = Tuple[Optional[Type[InferenceContext]], ...]
+
+
+class InferenceRule:
+    """Implements some sort of heuristic that produces signals about columns.
+
+    An inference rule produces zero or more `InferenceSignal` instances about whatever
+    columns it thinks it should, based on input `InferenceContext`s.
+
+    Concrete implementations should aim to be short and modularized. It is preferred to
+    compose multiple small rules that each produce a signal type than to make one large
+    rule with complex logic to produce a bunch of signals.
+    """
+
+    _required_contexts: RequiredContextsType
+
+    def __init_subclass__(cls) -> None:
+        """Initializes `cls._required_contexts` from `cls.process` type annotations."""
+        parameters = inspect.signature(cls.process).parameters
+        context_types: List[Optional[Type[InferenceContext]]] = []
+        for param in list(parameters.values())[1:]:  # exclude self
+            if param.annotation == inspect.Parameter.empty:
+                logger.warning(
+                    f"`{cls.__name__}.process`'s parameters must all be type annotated. "
+                    "Unanottated types will be provided None contexts, which might trigger further runtime errors."
+                )
+                context_types.append(None)
+                continue
+
+            if isinstance(param.annotation, str):
+                annotation_str = param.annotation
+            else:
+                annotation_str = param.annotation.__name__
+
+            annotation_type = InferenceContext.known_subclasses.get(annotation_str, None)
+            if annotation_type is None:
+                logger.warning(
+                    "`InferenceRule.process` arguments must be subclasses of `InferenceContext`. "
+                    "Some calls may have None arguments, which may result in runtime errors."
+                )
+            context_types.append(annotation_type)
+
+        cls._required_contexts = tuple(context_types)
+
+        super().__init_subclass__()
+
+    @property
+    def required_contexts(self) -> RequiredContextsType:
+        """The context types this rule requires to run successfully."""
+        return self._required_contexts
+
+    # NOTE: this is not an @abstractmethod since we don't expect children to implement the same signature
+    def process(self) -> List[InferenceSignal]:
```

Comment:
I like making the superclass infer the `required_contexts` from class annotations because it makes subclassing very easy and idiomatic, as in the PR description. However, this violates the principle that subclasses must keep the subclass interface. 

I wanna know your opinions on this. Maybe this is just too much magic? I kinda like it :)

Reply:
Btw, MyPy complains about this, so you need to add `# type: ignore` to each subclass implementation (which I hate). I haven't found a solution for this, though...